partners_api/taxi_engine.cpp

```diff
@@ -0,0 +1,203 @@
+#include "partners_api/taxi_engine.hpp"
+#include "partners_api/uber_api.hpp"
+#include "partners_api/yandex_api.hpp"
+
+#include "base/macros.hpp"
+#include "base/stl_add.hpp"
+
+#include <algorithm>
+#include <initializer_list>
+#include <iterator>
+
+namespace
+{
+template <typename Iter>
+Iter FindByProviderType(taxi::Provider::Type type, Iter first, Iter last)
+{
+  using IterValueType = typename std::iterator_traits<Iter>::value_type;
+  return std::find_if(first, last,
+                      [type](IterValueType const & item) { return item.m_type == type; });
+}
+}  // namespace
+
+namespace taxi
+{
+void ResultMaker::Reset(uint64_t requestId, size_t requestsCount,
+                        SuccessfullCallback const & successCallback,
+                        ErrorCallback const & errorCallback)
+{
+  std::lock_guard<std::mutex> lock(m_mutex);
+
+  m_requestId = requestId;
+  m_requestsCount = static_cast<uint8_t>(requestsCount);
+  m_successCallback = successCallback;
+  m_errorCallback = errorCallback;
+  m_providers.clear();
+  m_errors.clear();
+}
+
+void ResultMaker::DecrementRequestCount(uint64_t requestId)
+{
+  std::lock_guard<std::mutex> lock(m_mutex);
+
+  if (m_requestId != requestId)
+    return;
+
+  DecrementRequestCount();
+}
+
+void ResultMaker::ProcessProducts(uint64_t requestId, Provider::Type type,
+                                  std::vector<Product> const & products)
+{
+  std::lock_guard<std::mutex> lock(m_mutex);
+
+  if (m_requestId != requestId)
+    return;
+
+  m_providers.emplace_back(type, products);
+
+  DecrementRequestCount();
+}
+
+void ResultMaker::ProcessError(uint64_t requestId, Provider::Type type, ErrorCode code)
+{
+  std::lock_guard<std::mutex> lock(m_mutex);
+
+  if (m_requestId != requestId)
```

Comment:
Could you please explain how it's possible that requestId may be different from m_requestId? Is it possible to do this check in other place?